Compare window desktop id with currently active work area desktop id.

vldmr11080 opened the pull request:
<!-- Enter a brief description/summary of your PR here. What does it fix/what does it change/how was it tested (even manually, if necessary)? -->
## Summary of the Pull Request
Since handling of virtual desktop switch is asynchronous, event indicating that new window is created can appear in the meantime resulting in window being assigned to invalid zone. Compare virtual desktop id on which window is created with the currently active virtual desktop id, and continue handling window created event only in case these two match.

<!-- Please review the items on the PR checklist before submitting-->
## PR Checklist
* [x] Applies to #1236

<!-- Describe how you validated the behavior. Add automated tests wherever possible, but list manual validation steps taken as well -->
## Validation Steps Performed
1. Make sure you have `Move newly created windows to their last known zone` on.
2. Create a custom layout with two zones and apply it to both virtual desktops.
3. On first virtual desktop snap Notepad on zone 1.
4. On second virtual desktop snap Notepad on zone 2.
4. Switch back and forth between virtual desktops.

Expected result: Both windows stay assigned to their zones on different virtual desktops.

vldmr11080 commented:
@stefansjfw @enricogior please take a look, I rearranged code a bit and improved error handling.

stefansjfw commented on src/modules/fancyzones/lib/FancyZones.cpp:
```diff
@@ -349,17 +349,26 @@ FancyZones::VirtualDesktopInitialize() noexcept
         {
+            GUID windowDesktopId{};
+            GUID zoneWindowDesktopId{};
+            if (ZoneWindowUtils::GetWindowDesktopId(window, &windowDesktopId) &&
```
Maybe to call these 2 before if ?

yuyoyuppe commented on src/modules/fancyzones/lib/ZoneWindow.cpp:
```diff
@@ -81,6 +81,50 @@ namespace ZoneWindowUtils
+    IServiceProvider* GetServiceProvider()
+    {
+        static winrt::com_ptr<IServiceProvider> provider;
+        if (!provider)
```
this is not thread-safe. let's do it the same way as in `is_process_elevated`

yuyoyuppe commented on src/modules/fancyzones/lib/ZoneWindow.cpp:
```diff
@@ -81,6 +81,50 @@ namespace ZoneWindowUtils
+
+    IVirtualDesktopManager* GetVirtualDesktopManager()
+    {
+        static winrt::com_ptr<IVirtualDesktopManager> manager;
```
same here.

enricogior commented on src/modules/fancyzones/lib/ZoneWindow.cpp:
```diff
@@ -81,6 +81,44 @@ namespace ZoneWindowUtils
+
+    bool GetWindowDesktopId(HWND topLevelWindow, GUID* desktopId)
+    {
+        static IVirtualDesktopManager* virtualDesktopManager = GetVirtualDesktopManager();
```
We should only initialize the service provider and the `VirtualDesktopManager` once at FZ startup.
Move this to a different file, create a singleton that initialize the method, taking in consideration it may fail, in that case simple ignore the failure and always return nullptr in `GetWindowDesktopId`

vldmr11080 replied:
But wouldn’t this initialize it only once as it is declared as static according to C++ standard. Also it will guarantee that initialization is thread safe.

enricogior replied:
Sorry, it's late and I didn't see it. OK good.
Please move it outside of this file.

vldmr11080 replied:
Moved to `VirtualDesktopUtils.h(.cpp)`

vldmr11080 commented:
I've added additional check for **empty** GUID in `ZoneWindow` and moved helper functions around a bit. @PrzemyslawTusinski @stefansjfw @yuyoyuppe @enricogior could You review additional changes, and re-test again. Also have one test with multiple virtual desktops and one test where system is just started and only has one desktop with `00000000-0000-0000-0000-000000000000` GUID.

enricogior commented:
While testing this, we found out a very important piece of information: `virtualDesktopManager->GetWindowDesktopId` returns the real GUID for the primary desktop even when `RegistryHelpers::GetCurrentVirtualDesktop` returns the GUID with all zeros.
This requires some extra testing for this PR, but it will allow us to fix  (in a followup PR) the annoying zeroed GUID.

enricogior commented on src/modules/fancyzones/lib/FancyZones.cpp:
```diff
@@ -349,17 +350,26 @@ FancyZones::VirtualDesktopInitialize() noexcept
     {
         for (const auto& [monitor, zoneWindow] : m_zoneWindowMap)
         {
+            GUID windowDesktopId{};
```
Please add a comment explaining the context for this fix.
```
// WindowCreated is also invoked when a virtual desktop switch occurs, we need a way
// to figure out when that happens to avoid moving windows that should not be moved
```

vldmr11080 replied:
Comment added.
